text/3714-macro-fragment-fields.md

```diff
@@ -0,0 +1,163 @@
+[motivation]: #motivation
+
+The macros-by-example system is powerful, but sometimes difficult to work with.
+In particular, parsing complex parts of Rust syntax often requires carefully
+recreating large chunks of the Rust grammar, in order to parse out the desired
+pieces. Missing or incorrectly handling any portion of the syntax can result in
+not accepting the same syntax Rust does; this includes future extensions to
+Rust syntax that the macro was not yet aware of. Higher-level fragment
+specifiers are more robust for these cases, but don't allow extracting
+individual pieces of the matched syntax.
+
+This RFC introduces a mechanism to use high-level fragment specifiers while
+still extracting individual pieces of the matched syntax.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing macros by example, and using certain high-level fragment
+specifiers, you can use the syntax `${matched_name.field_name}` to extract
+specific "fields" of the matched syntax. This allows you to use the Rust parser
+for those high-level fragments, rather than having to recreate parts of the
+Rust grammar in order to extract the specific pieces you want. Fields evaluate
+to pieces of Rust syntax, suitable for substitution into the program or passing
+to other macros for further processing.
+
+For example, the fragment `:adt` parses any abstract data type supported by
+Rust: struct, union, or enum. Given a match `$t:adt`, you can obtain the name
+of the matched type with `${t.name}`:
+
+```rust
+macro_rules! get_name {
+    ($t:adt) => { println!("{}", stringify!(${t.name})); }
+}
+
+fn main() {
+    let n1 = get_name!(struct S { field: u32 });
+    let n2 = get_name!(enum E { V1, V2 = 42, V3(u8) });
+    let n3 = get_name!(union U { u: u32, f: f32 });
+    println!("{n3}{n1}{n2}"); // prints "USE"
+}
+```
+
+An attempt to access a field that doesn't exist will produce a compilation
+error on the macro definition, whether or not the specific macro rule gets
+invoked.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Fragment fields may be used in a macro transcriber anywhere the corresponding
+fragment name could be used. They must follow the same rules for repetition
+handling as the corresponding fragment (e.g. being used at the same level/kind
+of repetition).
+
+This RFC introduces the following new fragment specifiers, with specified fields:
+
+- `:fn`: A function item.
+  - `name`: The name of the function, as an `ident`.
+  - `params`: The parameters of the function, not including the surrounding
+    parentheses. May be empty if the function has no parameters.
+  - `return_type`: The return type of the function, as a `ty`. If the function
+    has no explicitly specified return type, this will be `()`, with a span of
+    the closing parenthesis for the function arguments.
+  - `body`: The body of the function, as a block (including the
+    surrounding braces).
```

Comment:
will this match fn declarations having no bodies in trait and extern blocks?

if yes, what is the content and span of the `.body`?

```rust
x! {
    extern "C" {
        fn foo();  // <-- will a `:fn` match this?
    }
}

x! {
    trait Tr {
        fn bar(&self); // <-- will a `:fn` match this?
        
        fn baz(&self, count: usize) {
            for _ in 0..count {
                self.bar();
            }
        }
    }
}
```

Reply:
@kennytm My intention is that `:fn` matches function *definitions* (including a body), not *declarations*. I've now clarified this.

I think if we want to match function *declarations* we should provide a separate specifier for that. I've added that to future work.